src/Symfony/Component/Console/Application.php

```diff
@@ -1004,34 +1004,51 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
             $commandSignals = $command instanceof SignalableCommandInterface ? $command->getSubscribedSignals() : [];
 
             if ($commandSignals || null !== $this->dispatcher) {
-                if (!$this->signalRegistry) {
+                if (!isset($this->signalRegistry)) {
                     throw new RuntimeException('Unable to subscribe to signal events. Make sure that the `pcntl` extension is installed and that "pcntl_*" functions are not disabled by your php.ini\'s "disable_functions" directive.');
                 }
 
                 if (Terminal::hasSttyAvailable()) {
                     $sttyMode = shell_exec('stty -g');
 
                     foreach ([\SIGINT, \SIGTERM] as $signal) {
-                        $this->signalRegistry->register($signal, static function () use ($sttyMode) {
-                            shell_exec('stty '.$sttyMode);
-                        });
+                        $this->signalRegistry->register($signal, static fn () => shell_exec('stty '.$sttyMode));
                     }
                 }
             }
 
+            $exitCode = 0;
```

Comment:
And I think I have something even more flexible:

```php
            if ($this->dispatcher) {
                // We register application signal, so that we can dispatch the event
                foreach ($this->signalsToDispatchEvent as $signal) {
                    $event = new ConsoleSignalEvent($command, $input, $output, $signal, 0);

                    $this->signalRegistry->register($signal, function ($signal) use ($event, $command, $commandSignals) {
                        $this->dispatcher->dispatch($event, ConsoleEvents::SIGNAL);
                        $exitCode = $event->getExitCode();

                        if (in_array($signal, $commandSignals)) {
                            $command->handleSignal($signal);
                            $exitCode = $command->getExitCode($signal, $exitCode);
                        }
                        if (null !== $exitCode) {
                            exit($exitCode);
                        }
                    });
                }

                // then we register command signal, but not if already handled by the dispatcher
                foreach (array_diff($this->signalsToDispatchEvent, $commandSignals) as $signal) {
                    $this->signalRegistry->register($signal, function (int $signal) use ($command): void {
                        $command->handleSignal($signal);
                        $exitCode = $command->getExitCode($signal, 0);
                        if (null !== $exitCode) {
                            exit($exitCode);
                        }
                    });
                }
            } else {
                foreach ($commandSignals as $signal) {
                    $this->signalRegistry->register($signal, function (int $signal) use ($command): void {
                        $command->handleSignal($signal);
                        $exitCode = $command->getExitCode($signal, 0);
                        if (null !== $exitCode) {
                            exit($exitCode);
                        }
                    });
                }
            }
        }
```



WDYT?

Reply:
Applied! Tests is comming (next week I guess)